### PR TITLE
[13.0][MIG] ebill_paynet

### PR DESCRIPTION
**Pull request by TDu**
This is based on #499 for 12.0 that has not been merged

- [x] account_invoice_export OCA/edi#212
- [x] base_ebill_payment_contract OCA/edi#207

### CODE REVIEW

**Comment by sebalix** [comment]
This depends on some other modules, maybe add that in `oca_dependencies.txt` to get tests running on Travis?

- `account_invoice_export` https://github.com/OCA/edi/pull/212
- `base_ebill_payment_contract` https://github.com/OCA/edi/pull/207

File: oca_dependencies.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+server-env
+queue
```
Comment (sebalix) [inline oca_dependencies.txt:2]: you will need `edi` too for https://github.com/OCA/edi/pull/212 and https://github.com/OCA/edi/pull/207.

File: ebill_paynet/components/api.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+SSL_TEST_CERTIFICATE = get_resource_path(
+    "ebill_paynet", "certificats", "test_services_chain.pem"
+)
+
+
+class PayNetDWS:
+    """PayNet DWS web services."""
+
+    def __init__(self, url, test_service):
+        settings = Settings(xml_huge_tree=True)
```
Comment (sebalix) [inline ebill_paynet/components/api.py]: I would use directly `zeep.Settings`, this avoid to read the imports statements to know what we are talking about as it's not an usual dependency in the Odoo ecosystem. But it's a matter of taste I suppose :)

File: ebill_paynet/components/api.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        session = Session()
```
Comment (sebalix) [inline ebill_paynet/components/api.py:29]: Same with `requests.Session` etc

File: ebill_paynet/data/transmit.method.xml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="paynet_transmit_method" model="transmit.method">
+        <field name="name">Paynet</field>
+        <field name="code">paynet</field>
+        <field name="customer_ok">True</field>
+        <!-- There is a bug on transmit_method module, needs to be set -->
+        <!-- to True so it is available for customer -->
+        <field name="supplier_ok">True</field>
```
Comment (sebalix) [inline ebill_paynet/data/transmit.method.xml]:
```suggestion
        <field name="supplier_ok" eval="True"/>
```

File: ebill_paynet/data/transmit.method.xml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="paynet_transmit_method" model="transmit.method">
+        <field name="name">Paynet</field>
+        <field name="code">paynet</field>
+        <field name="customer_ok">True</field>
```
Comment (sebalix) [inline ebill_paynet/data/transmit.method.xml]:
```suggestion
        <field name="customer_ok" eval="True"/>
```

File: ebill_paynet/models/paynet_invoice_message.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+from zeep.exceptions import Fault
+
+from odoo import fields, models
+from odoo.modules.module import get_module_root
+
+from ..components.api import PayNetDWS
+
+MODULE_PATH = get_module_root(os.path.dirname(__file__))
+INVOICE_TEMPLATE_2013 = "invoice-2013A.xml"
+INVOICE_TEMPLATE_2003 = "invoice-2003A.xml"
```
Comment (sebalix) [inline ebill_paynet/models/paynet_invoice_message.py:20]: As they are Jinja templates:
```suggestion
INVOICE_TEMPLATE_2013 = "invoice-2013A.xml.j2"
INVOICE_TEMPLATE_2003 = "invoice-2003A.xml.j2"
```
(files to rename accordingly)

File: ebill_paynet/models/paynet_invoice_message.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+TEMPLATE_DIR = MODULE_PATH + "/messages"
```
Comment (sebalix) [inline ebill_paynet/models/paynet_invoice_message.py]: Use `os`:
```suggestion
TEMPLATE_DIR = os.path.join(MODULE_PATH, "messages")
```

File: ebill_paynet/models/account_invoice.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+    _inherit = "account.move"
+
+    @api.onchange("transmit_method_id", "partner_id")
+    def _onchange_transmit_method_id(self):
+        if self.type not in ("out_invoice", "out_refund"):
+            return
+        paynet_method = self.env.ref("ebill_paynet.paynet_transmit_method")
+        if self.transmit_method_id == paynet_method:
+            # It is not called when selecting the customer ?
```
Comment (sebalix) [inline ebill_paynet/models/account_invoice.py]: Add a FIXME explaining it's due to the `transmit_method_id` added twice in `account_invoice_transmit_method`, it has to be fixed there

https://github.com/OCA/account-invoicing/pull/682/files?file-filters%5B%5D=.csv&file-filters%5B%5D=.html&file-filters%5B%5D=.png&file-filters%5B%5D=.py&file-filters%5B%5D=.rst&file-filters%5B%5D=.xml#r439977482

File: ebill_paynet/models/account_invoice.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        )
+        if not contract:
+            _logger.error(
+                "Paynet contract for {} not found for {}".format(
+                    partner.name, transmit_method.name
+                )
+            )
+        return contract
+
+    def create_paynet_message(self):
```
Comment (sebalix) [inline ebill_paynet/models/account_invoice.py:43]: It has to be a private method as you are returning a recordset (not compatible with the JSON/XML-RPC layer):
```suggestion
    def _create_paynet_message(self):
```
(or return the ID of the message record but you'll have to rebrowse it after)

File: ebill_paynet/models/account_invoice.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+        """ """
+        activity_type = "ebill_paynet.mail_activity_dws_error"
+        activity = self.activity_reschedule(
+            [activity_type], date_deadline=fields.Date.today()
+        )
+        values = {}
+        if not activity:
+            message = self.env.ref("ebill_paynet.dws_reject_invoice").render(
+                values=values
+            )
```
Comment (sebalix) [inline ebill_paynet/models/account_invoice.py:114]: `values` is empty here, is it now useless?

File: ebill_paynet/models/paynet_invoice_message.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+    _description = "Paynet shipment send to service"
+
+    service_id = fields.Many2one(
+        comodel_name="paynet.service",
+        string="Paynet Service",
+        required=True,
+        ondelete="restrict",
+        readonly=True,
+    )
+    invoice_id = fields.Many2one(comodel_name="account.move", ondelete="restrict")
```
Comment (sebalix) [inline ebill_paynet/models/paynet_invoice_message.py:37]: `required=True`?

File: ebill_paynet/models/paynet_invoice_message.py [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+    state = fields.Selection(
+        selection=[
+            ("draft", "Draft"),
+            ("sent", "Sent"),
+            ("done", "Done"),
+            ("reject", "Reject"),
+            ("error", "Error"),
+        ],
+        default="draft",
+    )
```
Comment (sebalix) [inline ebill_paynet/models/paynet_invoice_message.py:48]: `required=True`

File: ebill_paynet/models/paynet_invoice_message.py [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+    ic_ref = fields.Char(
+        string="IC Ref", size=14, help="Document interchange reference"
+    )
+    # Set with invoice_id.number but also with returned data from server ?
+    ref = fields.Char("Reference N°", size=35)
+    ebill_account_number = fields.Char("Paynet Id", size=20)
```
Comment (sebalix) [inline ebill_paynet/models/paynet_invoice_message.py:54]: `required=True`?

File: ebill_paynet/models/paynet_invoice_message.py [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+
+    @staticmethod
+    def format_date(date_string=None):
+        if not date_string:
+            date_string = datetime.now()
+        return date_string.strftime("%Y%m%d")
+
+    def _generate_payload(self):
+        self.ensure_one()
+        assert self.state == "draft"
```
Comment (sebalix) [inline ebill_paynet/models/paynet_invoice_message.py]: A `UserError` is better here (or in `send_to_paynet` method and keep the assert here) as nothing prevent the user to run this (no filter on `draft` state).

File: ebill_paynet/models/paynet_invoice_message.py [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+        return "SA%012d" % self.id
+
+    def send_to_paynet(self):
+        for message in self:
+            message.payload = message._generate_payload()
+            try:
+                shipment_id = message.service_id.take_shipment(message.payload)
+                message.shipment_id = shipment_id
+                message.state = "sent"
+            except Fault as e:
```
Comment (sebalix) [inline ebill_paynet/models/paynet_invoice_message.py:74]:
```suggestion
            except zeep.exceptions.Fault as e:
```

File: ebill_paynet/models/paynet_service.py [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+# The state for already acknowledge ones ArrivedAtDestination
+
+_logger = logging.getLogger(__name__)
+
+
+class PaynetService(models.Model):
+    _name = "paynet.service"
+    _description = "Paynet service configuration"
+
+    name = fields.Char()
```
Comment (sebalix) [inline ebill_paynet/models/paynet_service.py]: Having the name required is a good thing I think

File: ebill_paynet/models/paynet_service.py [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+    client_pid = fields.Char(string="Paynet ID", size=17, required=True)
+    use_test_service = fields.Boolean(string="Testing", help="Target the test service")
+    service_type = fields.Selection(
+        selection=[("b2b", "B2B"), ("b2c", "B2C")],
+        string="Service type",
+        default="b2b",
+        help="Specify the type of XML exchange with the service.",
+    )
+    partner_bank_id = fields.Many2one(
+        comodel_name="res.partner.bank", string="Bank account",
```
Comment (sebalix) [inline ebill_paynet/models/paynet_service.py]: `ondelete="restrict"`?

File: ebill_paynet/models/paynet_service.py [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+    def get_shipment_content(self, shipment_id):
+        """ """
+        self.ensure_one()
+        dws = PayNetDWS(self.url, self.use_test_service)
+        try:
+            res = dws.client.service.getShipmentContent(
+                Authorization=dws.authorization(self.username, self.password),
+                ShipmentID=shipment_id,
+            )
+        except Fault as e:
```
Comment (sebalix) [inline ebill_paynet/models/paynet_service.py:111]:
```suggestion
        except zeep.exceptions.Fault as e:
```

File: ebill_paynet/readme/INSTALL.rst [hunk 19]
```diff
@@ -0,0 +1,2 @@
+This module depends on `base_ebill_payment_contract` which is located on `account_invoicing`, so the OCA repository `account_invoicing` must also be present on the system.
```
Comment (sebalix) [inline ebill_paynet/readme/INSTALL.rst]:
```suggestion
This module depends on `base_ebill_payment_contract` which is located on `OCA/account-invoicing`, so this repository must also be present on the system.
```

File: ebill_paynet/views/ebill_payment_contract.xml [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+                <field
+                    name="paynet_account_number"
+                    attrs="{'invisible': [('is_paynet_contract', '=', False)]}"
+                />
+                <field
+                    name="paynet_service_id"
+                    attrs="{'invisible': [('is_paynet_contract', '=', False)]}"
+                />
+                <field name="is_paynet_contract" invisible="1" />
+            </field>
```
Comment (sebalix) [inline ebill_paynet/views/ebill_payment_contract.xml]: I would put this in a separate group as `base_ebill_payment_contract` module could be used outside of Paynet context I suppose, so based on my comment on https://github.com/OCA/edi/pull/207#discussion_r475585357:
```suggestion
        <group name="base" position="after">
            <group name="paynet" position="after">
                <field
                    name="paynet_account_number"
                    attrs="{'invisible': [('is_paynet_contract', '=', False)]}"
                />
                <field
                    name="paynet_service_id"
                    attrs="{'invisible': [('is_paynet_contract', '=', False)]}"
                />
                <field name="is_paynet_contract" invisible="1" />
            </group>
        </group>
```

File: ebill_paynet/components/api.py (hunk 2 above)
Comment (TDu) [reply ebill_paynet/components/api.py]: Yes clearer, but with an `import zeep` are we not importing more than necessary and it would be bad performance wise ? (real question that I don't know the answer)

File: oca_dependencies.txt (hunk 1 above)
Comment (TDu) [reply oca_dependencies.txt:2]: Added

File: ebill_paynet/models/paynet_service.py (hunk 16 above)
Comment (TDu) [reply ebill_paynet/models/paynet_service.py]: Agree, done

File: ebill_paynet/models/paynet_service.py (hunk 17 above)
Comment (TDu) [reply ebill_paynet/models/paynet_service.py]: Yes, done

File: ebill_paynet/readme/INSTALL.rst (hunk 19 above)
Comment (TDu) [reply ebill_paynet/readme/INSTALL.rst]: Changed

File: ebill_paynet/views/ebill_payment_contract.xml (hunk 20 above)
Comment (TDu) [reply ebill_paynet/views/ebill_payment_contract.xml]: Much better done

File: ebill_paynet/__manifest__.py [hunk 21]
```diff
@@ -0,0 +1,31 @@
+# Copyright 2019-2020 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Ebill Paynet",
```
Comment (yvaucher) [inline ebill_paynet/__manifest__.py]:
```suggestion
    "name": "eBill Paynet",
```

File: ebill_paynet/readme/CONFIGURE.rst [hunk 22]
```diff
@@ -0,0 +1,14 @@
...
+To create a service you need to be registred with SIXT Paynet service. Then the configuration of the service can be done in `Accounting - Configuration - Payments - Paynet Service`
+
+## Configure the customers
+
+A customer that wants to receive his invoices through Paynet will also need to register with the service.
+In Odoo to enable the sending of invoices for a specific customer through Paynet, the transmit method must be set accordingly for that customer. This is done on the customer form in the tab `Sales & Purchases` section `Sales`.
+
+## Configure the contracts
+
+The contracts specific to ebilling are located in `Accounting - Customers - Ebill Payment Contract`
```
Comment (yvaucher) [inline ebill_paynet/readme/CONFIGURE.rst]:
```suggestion
The contracts specific to e-billing are located in `Accounting - Customers - eBill Payment Contract`
```

File: ebill_paynet/views/message_template.xml [hunk 23]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <template id="dws_reject_invoice">
+        <div>
+            <p>The Paynet system rejected the ebill that was send.</p>
```
Comment (yvaucher) [inline ebill_paynet/views/message_template.xml]:
```suggestion
            <p>The Paynet system rejected the eBill that was send.</p>
```

File: ebill_paynet/views/message_template.xml [hunk 24]
```diff
@@ -0,0 +1,9 @@
...
+            <!-- <p><t t-esc="job_id" /></p> -->
```
Comment (yvaucher) [inline ebill_paynet/views/message_template.xml]: A left over?

**Review by yvaucher — CHANGES_REQUESTED** [review]
Small pre-review about terminology.

Please prefer `eBill` and `e-billing`

File: ebill_paynet/components/api.py (hunk 2 above)
Comment (yvaucher) [reply ebill_paynet/components/api.py]: @TDu good question to ask.

AFAICT `import` does 2 things, it imports the modules and submodules and fill the namespace.

The best practices says that `from module import *` is discouraged. Because it fills the namespace with every part of  `module`.

When using the `from` you target a submodule or code of the main module in any case you will also load the module itself.

If you test the following in your python shell to load :

```
>>> import sys
>>> sys.modules['zeep']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'zeep'
>>> from zeep import Settings
>>> Settings
<class 'zeep.settings.Settings'>
>>> sys.modules['zeep']
<module 'zeep' from '/home/yvaucher/.local/lib/python3.6/site-packages/zeep/__init__.py'>
>>> sys.modules['zeep'].Settings
<class 'zeep.settings.Settings'>
>>> sys.modules['zeep'].Client
<class 'zeep.client.Client'>
>>> Client
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'Client' is not defined
```

As you can see the `zeep` module is fully loaded but the namespace doesn't contains everything.



But what is a submodule?
https://docs.python.org/3/reference/import.html#submodules

A submodule is a sub file or directory that is not mentionned in the `__init__.py`
It's different from code that is part of the main module which must appear in the `__init__.py` file.



To deep in details, you can't import too much things unless you ask for a submodule you don't use.

Consider the following module with 2 submodules and some initialization code:

```
module
├── initialization.py
├── __init__.py
├── sub1.py
└── sub2.py
```

The content of `module/__init__.py` is 
``` python
from . import initialization

print('import whole module')
```

Now let's test it:

```
>>> import module
load initialization
import whole module
>>> module.initialization
<module 'module.initialization' from '/tmp/module/initialization.py'>
>>> module.sub1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'module' has no attribute 'sub1'
```
Only what is mentionned as `from . import ....`  (in `module/__init__.py` file) is loaded with the module.

-----------------
```
>>> from module import initialization
load initialization
import whole module
>>> from module import sub1
load sub1
>>> from module import sub1, sub2
load sub2
```
Same here, only what is mention as `from . import ....` (in `module/__init__.py` file) is loaded with the module.
You also see that it reminds what has been previously loaded.

-----------------
```
>>> from module import sub1
load initialization
import whole module
load sub1
```

When importing `sub1` the whole module is loaded anyway. But not other sub packages.

-----------------
```
>>> from module import *
load initialization
import whole module
>>> initialization
<module 'module.initialization' from '/tmp/module/initialization.py'>
>>> sub1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'sub1' is not defined
```
You might have expected as I did to see an import of sub1 and sub2 but it's not the case, which means that this is discouraged only due to the namespace filling.


Additional note: If you do want to load a subpackage which has a too generic name you could also define an alias on import using 
`from module import subpackage as msub`
This can save you from namespace conflicts.


In conclusion, the proposal of @sebalix is perfectly fine and has not impact other than having one different entry in the name space.


Some references:

https://docs.python.org/3/reference/import.html
https://docs.python.org/3/library/functions.html#__import__
https://stackoverflow.com/questions/12229580/python-importing-a-sub-package-or-sub-module

File: ebill_paynet/data/transmit.method.xml (hunk 4 above)
Comment (TDu) [reply ebill_paynet/data/transmit.method.xml]: Changed, thanks

File: ebill_paynet/__manifest__.py [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+    "version": "13.0.1.0.0",
+    "license": "AGPL-3",
+    "author": "Camptocamp SA,Odoo Community Association (OCA)",
+    "website": "https://github.com/OCA/l10n-switzerland",
+    "depends": [
+        "account",
+        "account_invoice_export",
+        "base_ebill_payment_contract",
+        "l10n_ch_base_bank",
+        "l10n_ch_qr",
```
Comment (yvaucher) [inline ebill_paynet/__manifest__.py]: This module has been renamed `l10n_ch_qriban`

File: ebill_paynet/models/paynet_invoice_message.py [hunk 26]
```diff
@@ -112,6 +112,7 @@ def _generate_payload(self):
         params = {
             "client_pid": self.service_id.client_pid,
             "invoice": self.invoice_id,
+            "invoice_line": self.invoice_id.paynet_invoice_line_ids(),
```
Comment (sebalix) [inline ebill_paynet/models/paynet_invoice_message.py]: I suggest `invoice_lines`?

Comment (TDu) [reply ebill_paynet/models/paynet_invoice_message.py]: Yes, fixed in next commit

**Comment by TDu** [comment]
I open a new PR https://github.com/OCA/l10n-switzerland/pull/585 see why in the description.
Thanks for all your reviews.

**Comment by TDu** [comment]
Replaced by #585 and #586.
All comments have been taken care of.
